### PR TITLE
Initial support for list filtering

### PR DESCRIPTION
**Pull request by gj**

### CODE REVIEW

File: models/oso.py [hunk 1]
```diff
@@ -75,12 +78,36 @@ def authorize(self, user, action, resource):
             "env",
         )
         self.oso.register_constant(dict(self._context), "context")
-        return self.oso.is_allowed(user, action, resource)
+
+        # TODO(ap): Should we handle model-level access using partials, too?
+        if isinstance(resource, str):
```
Comment (samscott89) [inline models/oso.py:83]: Follow up: we can make the model access method call this directly.

File: models/oso.py [hunk 2]
```diff
@@ -75,12 +78,36 @@ def authorize(self, user, action, resource):
...
+            return next(
+                self.oso.query_rule("allow_model", user, action, resource), False
```
Comment (samscott89) [inline models/oso.py:85]: Wait, how have we been heavily using iterators/generators for 9 months and this is the first I'm seeing this `, False` option :O

File: models/oso.py [hunk 3]
```diff
@@ -75,12 +78,36 @@ def authorize(self, user, action, resource):
...
+                    domain = expr
+            if domain == []:
+                domain = domain_expression.FALSE_DOMAIN
+            return domain
 
 
 class OsoUser(models.AbstractModel):
     _inherit = "res.users"
 
+    def _register_hook(self):
```
Comment (samscott89) [inline models/oso.py:108]: Why do we need these?

File: models/partial.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+from odoo.models import BaseModel
+from odoo.osv import expression as domain_expression
+
+from polar import Expression, Variable
+from polar.exceptions import UnsupportedError
+
+
+def make_comparison(op):
+    def compare(left, right):
+        # Odoo likes to operate on record IDs instead of records.
```
Comment (samscott89) [inline models/partial.py:13]: I bet this comment cost many, many hours

File: models/oso.py (hunk 3 above)
Comment (gj) [reply models/oso.py:108]: - `Oso._register_hook()` (re-)loads policies when oso is initialized.
- `OsoUser._register_hook()` calls `OsoBase._register_hook()` so as to not duplicate its logic.
- `OsoBase._register_hook()` registers new models with oso.

Outstanding question: are policies (re-)loaded upon registering a new module with Odoo? We think the answer is no.

File: models/oso.py (hunk 1 above)
Comment (gj) [reply models/oso.py]: 8cd72f88426aa865b2d6fc1aec6ff952aba291f5

**Review by samscott89 — APPROVED** [review]

**Pull request merged**